docs/fsharp/style-guide/conventions.md

```diff
@@ -0,0 +1,455 @@
+
+```fsharp
+// Good!
+namespace MyCode
+
+type MyClass() =
+    ...
+```
+
+Using a top-level module may not appear different when called only from F#, but should your code change to be consumed by C#, callers may be surprised.
+
+```fsharp
+// Bad!
+module MyCode
+
+type MyClass() =
+    ...
+```
+
+### Only use `[<AutoOpen>]` for nested private modules
+
+The `[<AutoOpen>]` can pollute callers and the answer to where something comes from is "magic". The only exception to this rule is the F# Core Library itself (though this fact is also a bit controversial).
+
+However, it is a great convenience if you have more complicated helper functionality for a public API that you wish to organize separately from that public API.
+
+```fsharp
+module MyAPI =
+    [<AutoOpen>]
+    module private Helpers =
+        let helper1 x y z =
+            ...
+
+
+    let myFunction1 x =
+        let y = ...
+        let z = ...
+
+        helper1 x y z
+```
+
+This lets you cleanly separate implementation details from the public API of a function without having to fully qualify a helper each time you call it.
+
+### Use `[<RequireQualifiedAccess>]` whenever names could conflict or you feel it helps with readability
+
+Adding the `[<RequireQualifiedAccess>]` attribute to a module indicates that the module may not be opened and that references to the elements of the module require explicit qualified access. For example, the Microsoft.FSharp.Collections.List module has this attribute.
+
+This is useful when functions and values in the module have names that are likely to conflict with names in other modules. Requiring qualified access can greatly increase the long-term maintainability and evolvability of a library.
+
+```fsharp
+[<RequireQualifiedAccess>]
+module StringHelpers =
+    let parse s = ...
+
+...
+
+let s = getAString()
+let parsed = StringHepers.parse s // Must qualify to use 'parse'
+```
+
+### Do not sort `open` statements alphanumerically
+
+Although it is possible to sort `open` statements, doing so could have confusing consequences. The order of declarations matters in F#, and this also includes `open` statements. Sorting them alphanumerically can shadow values, thus allowing you to interact with values which are likely not what you think they are.
+
+## Use classes to contain values which have side effects
+
+There are many times when initializing a value can have side effects, such as instantiating a context to a database or other remote resource. It is tempting to initialize such things in a module and use it in subsequent functions:
+
+```fsharp
+// This is bad!
+module MyApi =
+    let dep1 = File.ReadAllText "/Users/{your name}/connectionstring.txt"
+    let dep2 = Environment.GetEnvironmentVariable "DEP_2"
+    let dep3 = Random().Next()
+ 
+    let function1 arg = doStuffWith dep1 dep2 dep3 arg
+    let function2 arg = doSutffWith dep1 dep2 dep3 arg
```

Comment:
Few other notes around module design:
* Public module functions should rarely accept more than 3-4 arguments.
* Public module functions should not accept too many function arguments. If they do expose function arguments, their intent should be evident simply by examining the function signature (think hoogle). I cannot understand what this function does : `val myCoolFunc : (int -> bool -> string -> int) -> (int -> 'T -> string) -> string * 'T list`.
* Mostly avoid exposing third-order functions or higher, e.g. `val foo : ((string -> unit) -> string) -> string`.
* If your function still requires too many function arguments, consider refactoring so that it accepts an interface. This usually records intention better.
* If your functions take more than 4-5 arguments, consider converting your module into a static facade, exposing methods that name their individual parameters. Do make use of optional parameters when appropriate.
* Avoid mixing tupled arguments with curried style, e.g. `val foo : (int * string) -> int -> (bool * bool * bool) -> int`.
* Avoid using tuples as return types. Pairs and triples are occasionally acceptable, when the API clearly communicates the purpose of each element type. In most cases however, it might make sense to write a custom record type (or use anonymous records in future F# releases) that clearly labels each of the values.